kube-core/src/cel.rs

```diff
@@ -0,0 +1,282 @@
+    }
+}
+
+/// Validate takes schema and applies a set of validation rules to it. The rules are stored
+/// on the top level under the "x-kubernetes-validations".
+///
+/// ```rust
+/// use schemars::schema::Schema;
+/// use kube::core::{Rule, Reason, Message, validate};
+///
+/// let mut schema = Schema::Object(Default::default());
+/// let rules = vec![Rule{
+///     rule: "self.spec.host == self.url.host".into(),
+///     message: Some("must be a URL with the host matching spec.host".into()),
+///     field_path: Some("spec.host".into()),
+///     ..Default::default()
+/// }];
+/// validate(&mut schema, rules)?;
+/// assert_eq!(
+///     serde_json::to_string(&schema).unwrap(),
+///     r#"{"x-kubernetes-validations":[{"fieldPath":"spec.host","message":"must be a URL with the host matching spec.host","rule":"self.spec.host == self.url.host"}]}"#,
+/// );
+/// # Ok::<(), serde_json::Error>(())
+///```
+#[cfg(feature = "schema")]
+#[cfg_attr(docsrs, doc(cfg(feature = "schema")))]
+pub fn validate(s: &mut Schema, rules: Vec<Rule>) -> Result<(), serde_json::Error> {
+    match s {
+        Schema::Bool(_) => (),
+        Schema::Object(schema_object) => {
+            schema_object
+                .extensions
+                .insert("x-kubernetes-validations".into(), serde_json::to_value(rules)?);
+        }
+    };
+    Ok(())
+}
+
+/// Validate property mutates property under property_index of the schema
+/// with the provided set of validation rules.
+///
+/// ```rust
+/// use schemars::JsonSchema;
+/// use kube::core::{Rule, validate_property};
+///
+/// #[derive(JsonSchema)]
+/// struct MyStruct {
+///     field: Option<String>,
+/// }
+///
+/// let gen = &mut schemars::gen::SchemaSettings::openapi3().into_generator();
+/// let mut schema = MyStruct::json_schema(gen);
+/// let rules = vec![Rule::new("self != oldSelf")];
+/// validate_property(&mut schema, 0, rules)?;
+/// assert_eq!(
+///     serde_json::to_string(&schema).unwrap(),
+///     r#"{"type":"object","properties":{"field":{"type":"string","nullable":true,"x-kubernetes-validations":[{"rule":"self != oldSelf"}]}}}"#
+/// );
+/// # Ok::<(), serde_json::Error>(())
+///```
+#[cfg(feature = "schema")]
+#[cfg_attr(docsrs, doc(cfg(feature = "schema")))]
+pub fn validate_property(
+    s: &mut Schema,
+    property_index: usize,
+    rules: Vec<Rule>,
+) -> Result<(), serde_json::Error> {
+    match s {
+        Schema::Bool(_) => (),
+        Schema::Object(schema_object) => {
+            let obj = schema_object.object();
+            for (n, (_, schema)) in obj.properties.iter_mut().enumerate() {
+                if n == property_index {
+                    return validate(schema, rules);
+                }
+            }
+        }
+    };
+
+    Ok(())
+}
+
+/// Merge schema properties in order to pass overrides or extension properties from the other schema.
+///
+/// ```rust
+/// use schemars::JsonSchema;
+/// use kube::core::{Rule, merge_properties};
```

Comment:
presumably this is a doc test that actually runs?

Reply:
ah, yeah, i see it [running in ci](https://github.com/kube-rs/kube/actions/runs/12397366325/job/34607579384?pr=1649) 👍